[torch-ray] Step2 프로토콜과 XCTesT 학습

torch-ray opened the pull request:
## Step1 작업목록
- [x] 클래스 3단계 이상 구분
- [x] 하위 클래스 속성추가
- [x] 추가속성에 대한 인터페이스 작성
- [x] 자판기 구조체 생성
- [x] 모든 메서드 구현
- [x] 테스트코드 작성
- [x] 테스트코드 작성
- [ ] 세부 예시처럼 모든 테스트 케이스 작성

## 학습키워드
- Protocal과 SuperClass + XCTest
- 클래스간의 의존성을 버리기 위한 프로토콜의 학습 및 XCTest 작성

## 고민과 해결
- VendingMahcin은 왜 구조체로 만들어야 하는지 아직 이해가되지 않음
- Protocal을 활용하여 상속관계를 모두 없애는 것도 가능한데, 그러려면 Protocal에 프로퍼티를 생성해야함
- Protocal에 프로퍼티를 생성하는 것을 지양해야하는 이유는?
- SuperClass가 Protocal에 비해 가지는 장점은?

## 부연설명
- (생략)

## 질문거리
- Protocal에 프로퍼티를 생성하는 것을 지양해야하는 이유는 무엇일까요? 의존성이 높아지기 때문일까요? 아직까진 메서드 역시 의존성이 생기는 것으로 보여서 프로토콜에 메서드와 프로퍼티를 만드는 것의 차이점을 모르겠습니다...!

torch-ray committed:
Step1 간단한 피드백 내용 반영

torch-ray committed:
Milk, SoftDrink, Coffee Clss 생성

torch-ray committed:
Milk Class 프로퍼티 생성 및 초기화

농장코드를 반환하는 원산지를 알 수 있는 메서드 생성

torch-ray committed:
SoftDrink Class 프로퍼티 생성 및 초기화

제로 칼러리 여부를 반환하는 메서드 생성

torch-ray committed:
Coffee Class 프로퍼티 생성 및 초기화

디카페인 여부를 반환하는 메서드 생성

torch-ray committed:
StrawBeryyMilk Class , chocolateMilk Class 및 sellByDate 프로토콜 추가

sellByDate 프로토콜을 활용하여 3단계 SubClass의 객체들에게 채택하게 할 예정

torch-ray committed:
StrawBerry Class 및 Chocolate Class에 프로퍼티 생성 및 초기화

기존에 3단계 객체들에서 채택 예정이었던 sellByDate프로토콜을 2단계 객체들에서 채택하는 것으로 수정, 우선 Milk에서 채택하였고 이후 SoftDrink 및 Coffee에서 채택 예정

유통기한은 세부상품으로 나눠지기 전에 커피, 우유 등으로 나눠졌을 때 분류하는 것이 적절한 것으로 보임

torch-ray committed:
SellByDate 프로토콜 삭제

단순 프로퍼티 용도로 사용했던 유통기한 프로토콜을 삭제하고 상위클래스인 Milk의 프로퍼티로 선언, 초코우유나 딸기우유의 첨가물을 확인하는 메서드를 프로토콜로 구현 예정인데, additivie 프로퍼티가 공통적으로 사용되지만 상위 클래스인 Milk에는 필요가 없는 것 같아서 고민중

torch-ray committed:
3단계 하위 클래스 생성

torch-ray committed:
Coke Class프로퍼티 및 enum 타입 생성

torch-ray committed:
Sprite Class 프로퍼티 및 enum 타입 생성

torch-ray committed:
3단계 하위 클래스 top, contata 생성으로 마무리 및 유제품 첨가 여부 확인 프로토콜 생성

torch-ray committed:
DateFormatter 기능 별도로 메서드로 빼서 구현

Beverage Class에 있는게 맞는지 고민은 되지만, 해당 클래스 외에 적절한 위치를 찾지 못함

torch-ray committed:
VendingMachine 구조체 생성 및 음료수 입고 메서드 생성

torch-ray committed:
유통기한 프로퍼티 Beverage Class로 옮기고 isHot과 저칼로리 프로퍼티 추가

torch-ray committed:
CheckBeverageInfo 프로토콜 추가

해당 프로토콜 Beverage Class에서 채택예정

torch-ray committed:
Beverage Class에 프로토콜 채택하여 유통기한초과 여부, 음료 뜨거움, 저칼로리 여부 확인 메서드 구현

torch-ray committed:
VendingMachine 구조체에 음료를 추가하고, 돈을 넣고, 현재 구입가능한 음료 리스트를 리턴하는 메서드 구현

torch-ray committed:
beverageList 및 음료추가 메서드 구현

음료 추가 메서드는 추가할 음료의 수를 파라미터로 전달받으며 해당 파라미터만큼 for-loop로 beverage list를 탐색하며 음료수를 추가

torch-ray committed:
VendingMachine 구조체에 잔액, 재고, 뜨거운 음료 여부 확인할 수 있는 메서드 추가 완료

torch-ray committed:
Test Case 추가

XCTest를 활용한 테스트 코드를 추가하였으나, 아직 완벽하게 사용하는 방법을 모르겠어서 추가 학습이 필요함

godrm commented:
> * Protocal에 프로퍼티를 생성하는 것을 지양해야하는 이유는 무엇일까요? 의존성이 높아지기 때문일까요? 아직까진 메서드 역시 의존성이 생기는 것으로 보여서 프로토콜에 메서드와 프로퍼티를 만드는 것의 차이점을 모르겠습니다...!

프로토콜에 프로터피를 선언하면 값을 가져가는 방식으로 구현하게 되서 그렇고, private으로 할 수 없어서 처음부터 그걸 고려하면 객체 선언을 연습하기에 부적절하더군요. 문법적으로 가능하지만 잘못 사용할 가능성이 높아집니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -2,27 +2,76 @@ import Foundation
     private var productName: String
-    private var manufactureDate: Date
+    private var manufacturedDay: Date
+    private var sellByDate: Date
```
타입에 Date가 있으니 속성 이름에는 manufacturedAt 정도로 표현하기도 합니다. 참고하세요.

torch-ray replied:
수정하였습니다. 추후 네이밍에도 반영하겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -2,27 +2,76 @@ import Foundation
+    private var manufacturedDay: Date
+    private var sellByDate: Date
+    private var lowCalories: Bool
+    private var isHot: Bool
```
이런 속성은 칼로리나 온도를 값으로 저장하고, 그 값이 기준값보다 높은지 낮은지 판단해주는 메소드를 제공하는 게 좋습니다. Bool 타입은 정말 true,false로만 나눠지는 것에 선언하세요.

torch-ray replied:
Int형으로 변경하였습니다. 저칼로리나 음료의 온도여부는 보통 매장에서 true/false로 구분한다고 생각했습니다.. 참고하겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -2,27 +2,76 @@ import Foundation
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyyMMDD"
+        return dateFormatter.string(from: date)
+    }
```
이런 메소드는 음료 속성과 관련이 없는 메소드니까 여기에 선언할 필요가 없습니다. 
어디에 선언하면 좋은지 고민해보세요. 
힌트를 주자면 foo(a) 형태 메소드는 a.foo() 형태로 만들수도 있습니다.

torch-ray replied:
extension활용했습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -2,27 +2,76 @@ import Foundation
+    }
+    
+    func checkPrice() -> Int {
+        return price
```
두 가지를 고려해보세요.
- checkPrice는 가격을 어떻게 한다는 것인지 정보를 주지 못합니다. check-보다 구체적인 동작을 표현해보세요.
- 값을 가져가지 말고 price를 가져가서 비교하는 동작을 여기서 메소드로 제공해보세요.

torch-ray replied:
메서드명을 변경하고 beverage class에서 직접 비교하여 Bool타입을 리턴하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -2,27 +2,76 @@ import Foundation
+    }
+    
+    func canbuy() -> String {
+        return productName
```
이 메소드는 역할이 뭔가요? 살 수 있는지 판단하는 메소드인가요? 아니면 제품 이름을 가져가기 위한건가요?
제품 이름은 왜 필요하죠?

torch-ray replied:
제품명을 비교하여 재고확인 및 현재 구매 가능한 음료 구매시도를 하였는지 확인하는 용도입니다. 메서드명 변경하였고 해당 클래스 안에서 비교하는 기능추가하여 Bool타입으로 반환하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -2,27 +2,76 @@ import Foundation
 
+extension Beverage: CheckBeverageInfo {
+    
+    func validate(with date: Date) -> Bool {
```
validate- 표현도 날짜를 뭘한다는 것인지 모르겠습니다. 조금더 구체적인 동작을 표현하면 좋겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -2,27 +2,76 @@ import Foundation
+        formatter.locale = Locale(identifier: "ko_KR")
+        let nowDate = formatter.date(from: dateToString(with: date))!
+        let sellDate = formatter.date(from: dateToString(with: sellByDate))!
+        switch nowDate.compare(sellDate) {
```
날짜를 비교해야 한다면 세부 동작을 Date 타입을 확장해서 분리해보세요.
이 메소드 자체는 음료 자체가 구현해야 하는게 아니라 Date에 있어도 될 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -2,27 +2,76 @@ import Foundation
+extension Beverage: Equatable {
+    
+    public static func == (lhs: Beverage, rhs: Beverage) -> Bool {
+        return lhs.canbuy() == rhs.canbuy()
```
같은 파일 내에서 extension 한 것은 lhs.xxx 에서 private 변수에 접근이 가능합니다. 
이거 때문에 canbuy를 선언한 거라면 불필요합니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverages.swift:
```diff
@@ -0,0 +1,34 @@
+import Foundation
+
+class Beverages {
+    var beverages: [Beverage]
```
Beverages 타입을 선언한 것은 좋습니다. 
내부 속성은 private var로 선언하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Cantata.swift:
```diff
@@ -0,0 +1,15 @@
+
+class Cantata: Coffee, IsDairyProducts {
+    
+    private var dairyProducts: Bool
```
dairyProducts 는 어떤 의미인가요? Bool 타입인데 어떤 속성인지 모르겠어요.

godrm replied:
이 부분은 수정이 반영 안된 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/CheckAdditive.swift:
```diff
@@ -0,0 +1,5 @@
+import Foundation
+
+protocol CheckAdditive {
+    func checkAdditive() -> String
```
프로토콜 이름과 메소드 이름이 동일한데 어떤 의도인지 모르겠습니다. 
제품을 더하고 문자열을 리턴하는 건가요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/CheckBeverageInfo.swift:
```diff
@@ -0,0 +1,7 @@
+import Foundation
+
+protocol CheckBeverageInfo {
+    func validate(with date: Date) -> Bool
+    func isHotBeverage() -> Bool
+    func isLowCalorie() -> Bool
```
프로토콜을 선언한 것은 좋습니다. 이 프로토콜은 어떤 동작을, 어떤 타입을 추상화한 걸까요? 
모든 음료에서 필요한거라면 Beverage에 있는 것과 프로토콜로 나눈 것은 어떤 차이가 있을까요? 
왜 이 프로토콜이 필요한지 모르겠습니다. 수퍼 클래스로 할 수 있는 것을 다른 프로토콜로 선언할 필요는 없으니까요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Coffee.swift:
```diff
@@ -0,0 +1,16 @@
+    
+    func discriminateDecaffeinated () -> String {
+        if decaffeinated { return "YES" }
+        else { return "NO" }
```
속성은 Bool 타입인데 판단해서 넘겨주는 메소드는 String 이네요. 이 메소드는 어떤 역할일까요? 
문자열은 기본 타입 중에 가장 무겁고 느린 타입입니다. 이 메소드 역할은 뭔가요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/IsDairyProducts.swift:
```diff
@@ -0,0 +1,5 @@
+import Foundation
+
+protocol IsDairyProducts {
+    func isDiaryProducts () -> Bool
```
프로토콜 이름과 메소드 이름이 같은 것은 좋은 접근은 아닌 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/SoftDrink.swift:
```diff
@@ -0,0 +1,16 @@
+    
+    func discriminateZeroCalories() -> String {
+        if zeroCalories { return "YES" }
+        else { return "NO" }
```
칼로리처럼 양이 있는 정보는 양을 속성으로 갖고 있고, 기준값에 따라서 판단해주는 메소드를 구현하는 게 좋습니다.
문자열로 리턴하는 것이 왜 필요한지는 잘 이해가 안됩니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,82 @@
+                        Coke(brand: .coca, volume: 250, price: 1500, productName: "코카콜라 제로", manufacturedDay: Date(), sellByDate: Date(), lowCalories: true, isHot: false, zeroCalories: true, flavor: .Original),
+                        Sprite(brand: .coca, volume: 190, price: 340, productName: "스프라이트", manufacturedDay: Date(), sellByDate: Date(), lowCalories: false, isHot: false, zeroCalories: false, container: .Can),
+                        Top(brand: .maxim, volume: 380, price: 990, productName: "맥심TOP", manufacturedDay: Date(), sellByDate: Date(), lowCalories: false, isHot: true, decaffeinated: false, dairyProducts: false),
+                        Cantata(brand: .lotte, volume: 500, price: 2000, productName: "칸타타콘트라베이스블랙", manufacturedDay: Date(), sellByDate: Date(), lowCalories: true, isHot: false, decaffeinated: true, dairyProducts: false)]
```
음료수 객체 인스턴스를 생성하는 게 필요하면 팩토리 패턴을 학습해보시고, 팩토리 객체를 구현해보세요.
객체 생성과 자판기 객체는 분리하는 게 더 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,82 @@
+import Foundation
+
+struct VendingMachine {
+    var insertedMoney: Int
```
insertedMoney 도 별도 타입으로 선언해보세요.

torch-ray replied:
별도로 선언했습니다. 타입을 별도로 생성하니 수정해야할 부분이 생각보다 많았습니다 ㅠㅠ

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,82 @@
+        for i in 0..<count {
+            let idx = i % beverageList.count
+            beverages.append(from: beverageList[idx])
+        }
```
beverages 타입이 있으니 한꺼번에 추가하는 것도 처리할 수 있지는 않을까요?
같은 종류의 인스턴스를 만들어도 동일한 클래스일 뿐이지, 인스턴스까지 같은 것을 넣어야 하는 것은 아닙니다. 
음료수가 여러개라면 인스턴스도 여러개가 생성되야 합니다.

torch-ray replied:
음료수 하나씩 인스턴스생성하도록 코드 수정하였습니다. 그렇게 되면 beverageList는 굳이 필요없을 것 같은데 추후 작업에도 필요없으면 삭제하겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,82 @@
+        var buyingProduct = [Beverage]()
+        beverages.forEachBeverage{
+            if $0.canbuy() == productName {
+                if $0.checkPrice() <= insertedMoney {
```
클로저로 구현해야 한다고 해서 모든 메소드를 클로저 하나로 처리해야 하는 것은 아닙니다.
음료 조건에 맞춰서 필터링하는 메소드도 개별적으로 구현하는 게 더 좋습니다.

torch-ray replied:
조건 하나 분리하여 새로운 메소드로 만들었습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,82 @@
+        return expired
+    }
+    
+    func isHotProducts() -> [Beverage] {
```
보통 has- is-를 붙이면 Bool을 리턴한다고 생각합니다. 배열을 리턴하고 있어서 어색한 부분이 있네요.

torch-ray replied:
메서드명 변경했습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,82 @@
+        return buyingProduct
+    }
+    
+    func checkBalance() -> Int{
+        return insertedMoney
+    }
+    
+    func checkStock() -> [String: Int] {
+        return beverages.countOfBeverages()
+    }
```
여기 메소드도 check- 보다 좀 더 구체적인 동사를 사용해보세요.

torch-ray replied:
메서드명 변경했습니다!

godrm reviewed (changes requested):
수고하셨습니다. 다음 피드백을 확인해서 개선해보세요.

torch-ray committed:
Naming수정 및 온도, 칼로리 Bool타입에서 Int타입으로 수정

torch-ray committed:
check- 네이밍 변경 및 비교 동작 Beverage Class 로 이동

checkPrcie() 함수의 이름을 availableForBeverage로 변경하고 price를 리턴하여 vendingMachine에서 값을 비교하던 코드를 Beverage에서 비교하여 Bool타입을 리턴하도록 수정

torch-ray committed:
cayBuy 메서드 이름 변경

기존에 재고확인 및 물품 구매시 producName을 통해 비교할때 사용하던 canBuy메서드 이름을 sellingBeverageList로 이름을 변경하고 productName을 리턴하지 않고 비교할 이름을 가져와서 비교 후 Bool타입을 리턴하도록 수정

torch-ray committed:
InsertedMoney Class 생성 및 메서드 구현

신규 타입 생성에 따른 기존 코드 수정

torch-ray committed:
addBeverage 메서드 수정

기존에 beverageList를 for-loop로 탐색하면서 음료 count를 받아서 추가했는데 1개씩 인스턴스를 생성해서 추가하도록 메서드 기능 변경

torch-ray committed:
음료 조건에 따라 필터링하는 메소드 개별적으로 구현

torch-ray committed:
메서드 이름 변경

torch-ray committed:
메서드명 변경

godrm commented on VendingMachineApp/VendingMachineApp/Model/Milk.swift:
```diff
@@ -4,9 +4,9 @@ class Milk: Beverage {
     private var farmCode: String
     
-    init(brand: Brand.Name, volume: Int, price: Int, productName: String, manufacturedDay: Date, sellByDate: Date, lowCalories: Bool, isHot: Bool, farmCode: String) {
+    init(brand: Brand.Name, volume: Int, price: Int, productName: String, manufacturedDay: Date, sellByDate: Date, calories: Int, temparatureOfBeverage: Int, farmCode: String) {
```
temparatureOfBeverage 대신 그냥 temparature라고 하세요.
어차피 모든 타입이 다 Beverage를 상속 받은거잖아요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -1,61 +1,65 @@
-            if $0.checkPrice() <= insertedMoney {
+            if $0.availableForBeverage() {
                 availableList.append($0)
             }
```
이런 메소드는 그냥 beverages.filterAvailables() 메소드를 호출하는 건 어떨까요? 
모든 것을 클로저로 호출할 필요는 없습니다.

godrm commented:
다음 단계를 하면서 마저 개선해보세요.
